party.py

```diff
@@ -0,0 +1,180 @@
+    "Address"
+    __name__ = 'party.address'
+
+    def match_with_amazon_data(self, address_data):
+        """
+        Match the current address with the address_record.
+        Match all the fields of the address, i.e., streets, city, subdivision
+        and country. For any deviation in any field, returns False.
+
+        :param address_data: Dictionary of address data from amazon
+        :return: True if address matches else False
+        """
+        Country = Pool().get('country.country')
+        Subdivision = Pool().get('country.subdivision')
+
+        # Find country and subdivision based on amazon data
+        country, = Country.search([
+            ('code', '=', address_data['CountryCode']['value'])
+        ], limit=1)
+        subdivision = Subdivision.search_using_amazon_state(
+            address_data['StateOrRegion']['value'], country
+        )
+
+        return all([
+            self.name == address_data['Name']['value'],
+            self.street == address_data['AddressLine1']['value'],
+            self.streetbis == (
+                address_data.get('AddressLine2') and
+                address_data['AddressLine2'].get('value') or None
+            ),
+            self.zip == address_data['PostalCode']['value'],
+            self.city == address_data['City']['value'],
+            self.country == country,
+            self.subdivision == subdivision,
+        ])
+
+    @classmethod
+    def find_or_create_for_party_using_amazon_data(cls, party, address_data):
+        """
+        Look for the address in tryton corresponding to the address_record.
+        If found, return the same else create a new one and return that.
+
+        :param party: Party active record
+        :param address_data: Dictionary of address data from amazon
+        :return: Active record of address created/found
+        """
+        for address in party.addresses:
+            if address.match_with_amazon_data(address_data):
+                break
+
+        else:
+            address = cls.create_for_party_using_amazon_data(
+                party, address_data
+            )
+
+        return address
+
+    @classmethod
+    def create_for_party_using_amazon_data(cls, party, address_data):
+        """
+        Create address from the address record given and link it to the
+        party.
+
+        :param party: Party active record
+        :param address_data: Dictionary of address data from amazon
+        :return: Active record of created address
+        """
+        Country = Pool().get('country.country')
+        Subdivision = Pool().get('country.subdivision')
+        ContactMechanism = Pool().get('party.contact_mechanism')
+
+        country, = Country.search([
+            ('code', '=', address_data['CountryCode']['value'])
+        ], limit=1)
+        subdivision = Subdivision.search_using_amazon_state(
+            address_data['StateOrRegion']['value'], country
+        )
+
+        address, = cls.create([{
```

Comment:
building the values dictionary is there in two places. can you merge it into one method and in active record pattern ?

My suggestion

```python

def get_address_from_amazon_data(cls, data):
    # find country
    # find state
    return Address(
        name=data['Name']['value'],
        ....
    )
```

you can then use that for matching and for creating new address for party ?